crates/recursor/src/recursor_dns_handle.rs

```diff
@@ -377,31 +362,55 @@ impl RecursorDnsHandle {
         &self,
         zone: Name,
         request_time: Instant,
-    ) -> Result<RecursorPool<TokioRuntimeProvider>, Error> {
+        mut depth: u8,
+    ) -> Result<(u8, RecursorPool<TokioRuntimeProvider>), Error> {
         // TODO: need to check TTLs here.
         if let Some(ns) = self.name_server_cache.lock().get_mut(&zone) {
-            return Ok(ns.clone());
+            debug!("returning cached pool for {zone}");
+            return Ok((depth, ns.clone()));
         };
 
+        depth += 1;
+        Error::recursion_exceeded(self.ns_recursion_limit, depth, &zone)?;
+
         let parent_zone = zone.base_name();
 
-        let nameserver_pool = if parent_zone.is_root() {
+        let (mut ns_depth, nameserver_pool) = if parent_zone.is_root() {
             debug!("using roots for {zone} nameservers");
-            self.roots.clone()
+            (depth, self.roots.clone())
         } else {
-            self.ns_pool_for_zone(parent_zone, request_time).await?
+            self.ns_pool_for_zone(parent_zone, request_time, depth)
+                .await?
         };
 
-        // TODO: check for cached ns pool for this zone
+        let mut lookup = Query::query(zone.clone(), RecordType::NS);
 
-        let lookup = Query::query(zone.clone(), RecordType::NS);
+        let response = loop {
+            ns_depth += 1;
 
-        let response = self
-            .lookup(lookup.clone(), nameserver_pool.clone(), request_time, false)
-            .await?;
+            Error::recursion_exceeded(self.ns_recursion_limit, ns_depth, &zone)?;
 
-        // let zone_nameservers = response.name_servers();
-        // let glue = response.additionals();
+            match self
+                .lookup(lookup.clone(), nameserver_pool.clone(), request_time, false)
+                .await
+            {
+                Ok(response) => break response,
+                Err(e) => match e.kind() {
+                    ErrorKind::Forward(name) => {
+                        // if we already had this name, don't try again
+                        if zone == name.name {
+                            debug!("zone previously searched for {}", name.name);
+                            return Err(e);
+                        };
+
+                        debug!("ns for {zone} forwarded to {} via SOA record", name.name);
+                        lookup = Query::query(name.name.clone(), RecordType::NS);
+                        continue;
+                    }
+                    _ => return Err(e),
+                },
```

Comment:
Nit: since the `Ok()` arm is `break`ing, I think you can take this code out of the `match`?

Reply:
I broke this out.  I also noticed a missing call to ns_pool_for_zone for the Forward error case - and added that back in.